File: pages/refactorings/replace-global-state-with-parameter/03-fresh-names/Before.hs
-- Number every variable in an expression tree, pre-order, with a
-- fresh-number supply.
module Before where

import Data.IORef (IORef, newIORef, readIORef, writeIORef)
import System.IO.Unsafe (unsafePerformIO)

data Expr = Var String | Lit Int | Add Expr Expr
  deriving (Eq, Show)

supply :: IORef Int
{-# NOINLINE supply #-}
supply = unsafePerformIO (newIORef 0) -- global fresh-number supply

number :: Expr -> IO Expr
number e = do
  writeIORef supply 0
  go e

go :: Expr -> IO Expr
go (Var name) = do
  n <- readIORef supply
  writeIORef supply (n + 1)
  pure (Var (name ++ show n))
go e@(Lit _) = pure e
go (Add l r) = do
  l' <- go l
  r' <- go r
  pure (Add l' r')

File: pages/refactorings/replace-global-state-with-parameter/03-fresh-names/Before.scala
// Number every variable in an expression tree, pre-order, with a
// fresh-number supply.
object Before:
  enum Expr:
    case Var(name: String)
    case Lit(value: Int)
    case Add(lhs: Expr, rhs: Expr)

  var supply = 0 // global fresh-number supply

  def number(e: Expr): Expr =
    supply = 0
    go(e)

  private def go(e: Expr): Expr = e match
    case Expr.Var(name) =>
      val n = supply
      supply += 1
      Expr.Var(name + n)
    case Expr.Lit(_) => e
    case Expr.Add(l, r) => Expr.Add(go(l), go(r))
